flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgram.java

```diff
@@ -0,0 +1,571 @@
+        double filterRatio = computeFilterRatio(buildSide, probeSide, buildIndices, probeIndices);
+
+        RelNode localBuilder =
+                new BatchPhysicalLocalRuntimeFilterBuilder(
+                        buildSide.getCluster(),
+                        buildSide.getTraitSet(),
+                        buildSide,
+                        buildIndices.toIntArray(),
+                        buildRowCount,
+                        maxRowCount);
+        RelNode globalBuilder =
+                new BatchPhysicalGlobalRuntimeFilterBuilder(
+                        localBuilder.getCluster(),
+                        localBuilder.getTraitSet(),
+                        createExchange(localBuilder, FlinkRelDistribution.SINGLETON()),
+                        maxRowCount);
+        RelNode runtimeFilter =
+                new BatchPhysicalRuntimeFilter(
+                        probeSide.getCluster(),
+                        probeSide.getTraitSet(),
+                        createExchange(globalBuilder, FlinkRelDistribution.BROADCAST_DISTRIBUTED()),
+                        probeSide,
+                        probeIndices.toIntArray(),
+                        filterRatio);
+
+        return runtimeFilter;
+    }
+
+    /**
+     * Find a suitable build side. In order not to affect MultiInput, when the original build side
+     * of runtime filter is not an {@link Exchange}, we need to push down the builder, until we find
+     * an exchange and inject the builder there.
+     *
+     * @param rel the original build side
+     * @param buildIndices build indices
+     * @param buildSideChecker check whether current build side is suitable
+     * @return An optional info of the suitable build side.It will be empty if we cannot find the
+     *     suitable build side.
+     */
+    private static Optional<BuildSideInfo> findSuitableBuildSide(
+            RelNode rel,
+            ImmutableIntList buildIndices,
+            BiFunction<RelNode, ImmutableIntList, Boolean> buildSideChecker) {
+        if (rel instanceof Exchange) {
+            Exchange exchange = (Exchange) rel;
+            if (!(exchange.getInput() instanceof BatchPhysicalRuntimeFilter)
+                    && buildSideChecker.apply(exchange.getInput(), buildIndices)) {
+                return Optional.of(new BuildSideInfo(exchange.getInput(), buildIndices));
+            }
+        } else if (rel instanceof BatchPhysicalRuntimeFilter) {
+            // runtime filter should not as build side
+            return Optional.empty();
+        } else if (rel instanceof Calc) {
+            Calc calc = ((Calc) rel);
+            RexProgram program = calc.getProgram();
+            List<RexNode> projects =
+                    program.getProjectList().stream()
+                            .map(program::expandLocalRef)
+                            .collect(Collectors.toList());
+            ImmutableIntList inputIndices = getInputIndices(projects, buildIndices);
```

Comment:
Can this method be reused with DPP utils?

Reply:
No, the logic is not exactly same. In runtime filter, the return value of `getInputIndices` is either an empty list(which means the push-down is failed) or a list of the same length as `buildIndices`(because it requires a one-to-one correspondence between the build-indices and probe-indices). But in DPP, it does not have this requirement.